index.js

```diff
@@ -1,201 +1,114 @@
+        if (!checkIfInitialized(RNOneSignal)) return;
+        isValidCallback(observer);
 
-        // Make native request to init IAM handler
-        if (type === IN_APP_MESSAGE_CLICKED_EVENT) {
-            if (Platform.OS === 'android') {
-                RNOneSignal.initInAppMessageClickHandlerParams();
-            } else if (Platform.OS === 'ios') {
-                RNOneSignal.setInAppMessageClickHandler();
-            }
+        if (Platform.OS === 'android') {
+            RNOneSignal.addSubscriptionObserver();
         }
+        eventManager.addEventHandler(SUBSCRIPTION_CHANGED, observer);
+    }
 
-        // Check if there is a cache for this type of event
-        var cache = _notificationCache.get(type);
-        if (handler && cache) {
-            handler(cache);
-            _notificationCache.delete(type);
+    static addEmailSubscriptionObserver(observer) {
+        if (!checkIfInitialized(RNOneSignal)) return;
+        isValidCallback(observer);
+
+        if (Platform.OS === 'android') {
+            RNOneSignal.addEmailSubscriptionObserver();
         }
+        eventManager.addEventHandler(EMAIL_SUBSCRIPTION_CHANGED, observer);
     }
 
-    static removeEventListener(type) {
-        if (!checkIfInitialized()) return;
+    /* H A N D L E R S */
 
-        invariant(
-            type === NOTIFICATION_RECEIVED_EVENT ||
-            type === NOTIFICATION_OPENED_EVENT ||
-            type === IDS_AVAILABLE_EVENT ||
-//            type === SUBSCRIPTION_EVENT ||
-//            type === PERMISSION_EVENT ||
-            type === EMAIL_SUBSCRIPTION_EVENT ||
-            type === IN_APP_MESSAGE_CLICKED_EVENT,
-            'OneSignal only supports received, opened, ids, emailSubscription, and inAppMessageClicked events'
-        );
-
-        _eventTypeHandler.delete(type);
+    /**
+     * Sets the handler that fires before the notification is displayed
+     * Callback parameter is a `NotificationReceivedEvent` with:
+     *  - notification data
+     *  - `complete` function that accepts the `NotificationReceivedEvent`
+     * @param  {function} handler
+     */
+    static setNotificationWillShowInForegroundHandler(handler){
+        if (!checkIfInitialized(RNOneSignal)) return;
+        isValidCallback(handler);
+        RNOneSignal.setNotificationWillShowInForegroundHandler();
+        eventManager.setEventHandler(NOTIFICATION_WILL_SHOW, handler);
     }
 
-    static clearListeners() {
-        if (!checkIfInitialized()) return;
+    static setNotificationOpenedHandler(handler){
+        if (!checkIfInitialized(RNOneSignal)) return;
+        isValidCallback(handler);
 
-        for(var i = 0; i < _eventNames.length; i++) {
-            _listeners[_eventNames].remove();
-        }
+        RNOneSignal.setNotificationOpenedHandler();
+        eventManager.setEventHandler(NOTIFICATION_OPENED, handler);
     }
 
+    /* R E G I S T R A T I O N  E T C */
+
     static registerForPushNotifications() {
```

Comment:
Remove `registerForPushNotifications` method, it does the same thing as `promptForPushNotificationsWithUserResponse`

Reply:
Done